src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs

```diff
@@ -46,96 +43,209 @@ public class Settings : ISettings
-        private Settings _next;
-        // The priority of this setting file
-        private int _priority;
+        private SettingsFile _settingsHead { get; }
 
-        private bool Cleared { get; set; }
+        private Dictionary<string, AbstractSettingSection> _computedSections { get; set; }
 
-        public Settings(string root /*, ILogger logger */)
-            : this(root, DefaultSettingsFileName, false)
+        public SettingSection GetSection(string sectionName)
         {
+            if (_computedSections.TryGetValue(sectionName, out var section))
+            {
+                return section.Clone() as SettingSection;
+            }
+
+            return null;
         }
 
-        public Settings(string root, string fileName)
-            : this(root, fileName, false)
+        public void AddOrUpdate(string sectionName, SettingItem item)
         {
+            if (string.IsNullOrEmpty(sectionName))
+            {
+                throw new ArgumentNullException(nameof(sectionName));
+            }
+
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            Debug.Assert(item.IsAbstract());
+
+            // Operation is an update
+            if (_computedSections.TryGetValue(sectionName, out var section) && section.Items.Contains(item))
+            {
+                // An update could not be possible here because the operation might be
+                // in a machine wide config. If so then we want to add the item to
+                // the output config.
+                if (section.Update(item))
+                {
+                    return;
+                }
+            }
+
+            // Operation is an add
+            var outputSettingsFile = GetOutputSettingFileForSection(sectionName);
+            if (outputSettingsFile != null)
+            {
+                AddOrUpdate(outputSettingsFile, sectionName, item);
+            }
         }
 
-        public Settings(string root, string fileName, bool isMachineWideSettings)
+        internal void AddOrUpdate(SettingsFile settingsFile, string sectionName, SettingItem item)
         {
-            if (string.IsNullOrEmpty(root))
+            if (string.IsNullOrEmpty(sectionName))
+            {
+                throw new ArgumentNullException(nameof(sectionName));
+            }
+
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            if (!Priority.Contains(settingsFile))
+            {
+                Priority.Last().SetNextFile(settingsFile);
+            }
+
+            Debug.Assert(item.IsAbstract());
+
+            settingsFile.AddOrUpdate(sectionName, item);
+
+            if (_computedSections.TryGetValue(sectionName, out var section))
             {
-                throw new ArgumentException("root cannot be null or empty", nameof(root));
+                section.Add(item);
```

Comment:
Since `_computedSection` only has pointers to the items that are held in the `settingsFile` the `AddOrUpdate` above this line takes care of the update scenario. I will add some comments to clarify this.